02-express-tutorial/app.js

```diff
@@ -1 +1,51 @@
-console.log('Express Tutorial')
+const express = require('express');
+const { products } = require('./data')
+
+const app = express();
+app.use(express.static('./public'))
+
+
+app.get('/api/v1/test', (req, res) => {
+    res.json({ message: 'It worked!' })
+});
+
+app.get('/api/v1/products', (req, res) => {
+    res.json(products)
+});
+
+app.get('/api/v1/products/:productID', (req, res) => {
+    const idToFind = parseInt(req.params.productID); 
+    const product = products.find((p) => p.id === idToFind);
+    if(idToFind === NaN || !product) {
+        return res.json({ message: "Unfortunately, that product was not found." })
+    }
+    return res.json(product);
+});
+
+app.get('/api/v1/query', (req, res) => {
+    const { search, limit } = req.query;
+    let searchedProducts = []; 
+    if(search) {
+        searchedProducts = products.filter((product) => {
+            return product.name.startsWith(search)
+        });
+    }
+
+    if(limit) {
+        searchedProducts = searchedProducts.slice(0, parseInt(limit))
+    }
+
+    if(!searchedProducts.length) {
+        return res.json({ message: 'Product you are searching not found' })
```

Comment:
# Required Improvement
[Status Code Conventions]
Again here if the response schema (aka: format, data model, etc) is different when no results are found, its not good to give a `200 Success` status code.
The convention or best practice is a little different for query requests though. `4XX` status codes like `404 Not Found` are used when the caller makes a mistake, such as requesting an ProductId that does not exist. In the case of a query, the caller has actually not made a mistake per se, they searched for something and they knew (or should have assumed) there was a chance that no results will match their query. So generally you'd an empty array.
```js
res.status(200).json(searchedProducts)
```

If you wanted to be a little more helpful to the user calling the API it might be better to return something like this instead:
```js
{
  "success": bool,
  "data": Array[Product],
  "total_count": int,
  "more": bool
}
```
Here the `data` is your `searchedProducts` array and the `success` is whether or not anything was found. Then the `total_count` tells them how many products in total matched their query (before the limit was applied) and the `more` tells them if the response contains all the results or if it was truncated (aka: cutoff) due to the limit.

If your code uses this response schema, it needs to use it in both places - when 0 results are found and when >0 are found. That was the caller knows what schema to expect at all times when a `200 Success` is returned.